smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+                .withMember(SHAPES_TEXT, shapesBuilder.build())
+                .build();
+    }
+
+    /**
+     * Finds invalid types and tags and either removes them or throws an error depending on
+     * whether toThrow is true or false.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in definitions.
+     */
+    public void validateTypesAndTags() {
+        //The optional ArtifactDefinitions must be non-null to call this method
+        Objects.requireNonNull(artifactDefinitions);
+
+        //for each entry in the shapes map
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            //for each ShapeLink in entry's List<ShapeLink>
+            for (ShapeLink link : entry.getValue()) {
+                //checking if link's type is in artifactDefinitions
+                if (!artifactDefinitions.getTypes().containsKey(link.getType())) {
+                    throw new ExpectationNotMetException(entry.getKey().toString()
+                            + " contains types that aren't in definitions.", sl);
+                }
+
+                //checking if link's tags are all in artifactDefinitions
+                Optional<List<String>> tags = link.getTags();
+                if (tags.isPresent()) {
+                    for (String tag : tags.get())
+                        if (!artifactDefinitions.getTags().containsKey(tag)) {
+                            throw new ExpectationNotMetException(entry.getKey().toString() + " " + tag
+                                    + " is a tag that isn't in definitions.", sl);
+                        }
+                }
+            }
+        }
+    }
+
+    /**
+     * Parses model and determines whether the trace file object meets the specs of the model by checking if
+     * the trace file contains all the ShapeIds in the model and the model contains all the ShapeIDs in the trace file.
+     *
+     * @param modelResourceName the model name to validate the trace file against. Model should be in the
+     *                          resources file.
+     * @throws ExpectationNotMetException if model contains a ShapeID not in TraceFile or TraceFile contains a ShapeID
+     *                                    not in model.
+     */
+    public void validateModel(String modelResourceName) {
+        Model model = Model.assembler()
+                .addImport(getClass().getResource(modelResourceName))
+                .assemble()
+                .unwrap();
+
+        //model contains all the shapeIds in shapes.keySet()
+        for (ShapeId id : shapes.keySet()) {
+            model.expectShape(id);
+        }
+
+        //shapes.keySet() contains all the shapeIds in model, we don't care about shapes in smithy.api namespace
+        for (Shape shape : model.toSet()) {
+            ShapeId id = shape.getId();
+            if (!id.getNamespace().equals("smithy.api") && !shapes.containsKey(id)) {
```

Comment:
This first check should probably be replaced with [`Prelude.isPublicPreludeShape`](https://awslabs.github.io/smithy/javadoc/1.0.6/software/amazon/smithy/model/loader/Prelude.html#isPublicPreludeShape(software.amazon.smithy.model.shapes.ToShapeId)).